// @azure/core-amqp/dist-esm/src/ConnectionContextBase.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { Connection, generate_uuid } from "rhea-promise";
import { CbsClient } from "./cbs";
import { ConnectionConfig } from "./connectionConfig/connectionConfig";
import { Constants } from "./util/constants";
import { getFrameworkInfo, getPlatformInfo } from "./util/runtimeInfo";
import { isNode } from "./util/utils";
// eslint-disable-next-line @typescript-eslint/no-redeclare -- renaming constant would be a breaking change.
export const ConnectionContextBase = {
    /**
     * Creates the base connection context.
     * @param parameters - Parameters to be provided to create
     * the base connection context.
     */
    create(parameters) {
        var _a, _b, _c;
        ConnectionConfig.validate(parameters.config, {
            isEntityPathRequired: parameters.isEntityPathRequired || false
        });
        const userAgent = parameters.connectionProperties.userAgent;
        if (userAgent.length > Constants.maxUserAgentLength) {
            throw new Error(`The user-agent string cannot be more than ${Constants.maxUserAgentLength} characters in length.` +
                `The given user-agent string is: ${userAgent} with length: ${userAgent.length}`);
        }
        const connectionOptions = {
            transport: Constants.TLS,
            host: parameters.config.host,
            hostname: (_a = parameters.config.amqpHostname) !== null && _a !== void 0 ? _a : parameters.config.host,
            username: parameters.config.sharedAccessKeyName,
            port: (_b = parameters.config.port) !== null && _b !== void 0 ? _b : 5671,
            reconnect: false,
            properties: {
                product: parameters.connectionProperties.product,
                version: parameters.connectionProperties.version,
                "user-agent": userAgent,
                platform: getPlatformInfo(),
                framework: getFrameworkInfo()
            },
            idle_time_out: Constants.defaultConnectionIdleTimeoutInMs,
            operationTimeoutInSeconds: parameters.operationTimeoutInMs
                ? parameters.operationTimeoutInMs / 1000
                : undefined
        };
        if (parameters.config.webSocket ||
            (!isNode && typeof self !== "undefined" && self.WebSocket)) {
            const socket = parameters.config.webSocket || self.WebSocket;
            const host = parameters.config.host;
            const endpoint = parameters.config.webSocketEndpointPath || "";
            const socketOptions = parameters.config.webSocketConstructorOptions || {};
            const port = (_c = parameters.config.port) !== null && _c !== void 0 ? _c : 443;
            connectionOptions.webSocketOptions = {
                webSocket: socket,
                url: `wss://${host}:${port}/${endpoint}`,
                protocol: ["AMQPWSB10"],
                options: socketOptions
            };
        }
        const connection = new Connection(connectionOptions);
        const connectionLock = `${Constants.establishConnection}-${generate_uuid()}`;
        const connectionContextBase = {
            wasConnectionCloseCalled: false,
            connectionLock: connectionLock,
            negotiateClaimLock: `${Constants.negotiateClaim}-${generate_uuid()}`,
            connection: connection,
            connectionId: connection.id,
            cbsSession: new CbsClient(connection, connectionLock),
            config: parameters.config,
            refreshConnection() {
                const newConnection = new Connection(connectionOptions);
                const newConnectionLock = `${Constants.establishConnection}-${generate_uuid()}`;
                this.wasConnectionCloseCalled = false;
                this.connectionLock = newConnectionLock;
                this.negotiateClaimLock = `${Constants.negotiateClaim} - ${generate_uuid()}`;
                this.connection = newConnection;
                this.connectionId = newConnection.id;
                this.cbsSession = new CbsClient(newConnection, newConnectionLock);
            }
        };
        return connectionContextBase;
    }
};
